### PR TITLE
feat(blockifier): add secp logic

### PR DESCRIPTION
**Pull request by xrvdg**
This PR adds the logic for secp in preparation of the native syscalls. 

The code has been refactored such that the implementation is shared between the native and non-native syscalls.

### CODE REVIEW

**Review by meship-starkware — COMMENTED** [review]
> Reviewed 3 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1 and @xrvdg)

**Comment by xrvdg** [comment]
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 450 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OB4nL7D3e7FKvXBKBHC:-OB9UEPn2e1CbYucT_AU:b68w4b) ([raw file](https://github.com/starkware-libs/sequencer/blob/5c58150e26378dc33ed5b2d113e77903cd78c595/crates/blockifier/src/execution/native/syscall_handler.rs#L450)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

Can we move these tests to the secp test? I am not sure this is the best solution, but I don't like the tests to be part of the syscall handler file.
</blockquote></details>

These can be moved to somewhere else, but that would require making `Secp256Point` public.

**Review by meship-starkware — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: all files reviewed, 1 unresolved discussion (waiting on @noaov1 and @xrvdg)

___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 450 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OB4nL7D3e7FKvXBKBHC:-OBLDOiw3_FjhHMKj69Q:b-kb5pnt) ([raw file](https://github.com/starkware-libs/sequencer/blob/5c58150e26378dc33ed5b2d113e77903cd78c595/crates/blockifier/src/execution/native/syscall_handler.rs#L450)):*
<details><summary><i>Previously, xrvdg (Xander van der Goot) wrote…</i></summary><blockquote>

These can be moved to somewhere else, but that would require making `Secp256Point` public.
</blockquote></details>

I see. Then I think it is better to keep it as is

**Review by meship-starkware — COMMENTED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: all files reviewed, 1 unresolved discussion (waiting on @noaov1 and @xrvdg)

**Review by rodrigo-pino — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @meship-starkware and @noaov1)

___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 436 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OB4n9eh1Eh9Vr36Q0BM:-OBPu59P9PrjQmGehy_w:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/5c58150e26378dc33ed5b2d113e77903cd78c595/crates/blockifier/src/execution/native/syscall_handler.rs#L436)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

Please avoid using super
</blockquote></details>

Done

**Review by meship-starkware — APPROVED** [review]
> Reviewed 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @noaov1)

**Review by meship-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @noaov1)

**Review by meship-starkware — APPROVED** [review]
> Reviewed 1 of 1 files at r13, all commit messages.
> Dismissed @graphite-app[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: all files reviewed, 1 unresolved discussion (waiting on @avi-starkware, @noaov1, @xrvdg, and @Yoni-Starkware)

**Comment by avi-starkware** [comment]
*[`crates/blockifier/src/execution/syscalls/secp.rs` line 40 at r13](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBaGRf74t_0VFCNyIvF:-OBaGRf74t_0VFCNyIvG:b-zdsgcx) ([raw file](https://github.com/starkware-libs/sequencer/blob/075467f2bce9281864e660c2cb349d2fc6cc2cfd/crates/blockifier/src/execution/syscalls/secp.rs#L40)):*
> ```XML
>     pub fn secp_mul(&mut self, request: SecpMulRequest) -> SyscallResult<SecpMulResponse> {
>         let ep_point = self.get_point_by_id(request.ec_point_id)?;
>         let result = *ep_point * Curve::ScalarField::from(request.multiplier);
> ```

@ilyalesokhin-starkware 
?



_Suggestion:_
```XML
        let ec_point = self.get_point_by_id(request.ec_point_id)?;
        let result = *ec_point * Curve::ScalarField::from(request.multiplier);
```

**Review by avi-starkware — COMMENTED** [review]
> Reviewed 2 of 4 files at r6, 1 of 1 files at r13.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1, @xrvdg, and @Yoni-Starkware)

**Review by avi-starkware — APPROVED** [review]
> Reviewed 1 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1, @xrvdg, and @Yoni-Starkware)

**Review by Yoni-Starkware — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: 3 of 4 files reviewed, 7 unresolved discussions (waiting on @avi-starkware, @meship-starkware, @noaov1, and @xrvdg)

___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 673 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfmlR99hFBCKV7rBdu:-OBfmlR99hFBCKV7rBdv:b-uswotf) ([raw file](https://github.com/starkware-libs/sequencer/blob/eb91092cf7161316374f56196a0996516cee1935/crates/blockifier/src/execution/native/syscall_handler.rs#L673)):*
> ```XML
>             Ok(None) => Ok(None),
>             Ok(Some(point)) => Ok(Some(Secp256Point(point))),
>             Err(error) => Err(error),
> ```

Please share this code



_Code quote:_
```XML
            Ok(None) => Ok(None),
            Ok(Some(point)) => Ok(Some(Secp256Point(point))),
            Err(error) => Err(error),
```
___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 679 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfmuYW0VBBCGj9xzDZ:-OBfmuYW0VBBCGj9xzD_:b-nz8x9i) ([raw file](https://github.com/starkware-libs/sequencer/blob/eb91092cf7161316374f56196a0996516cee1935/crates/blockifier/src/execution/native/syscall_handler.rs#L679)):*
> ```XML
> 
> /// Data structure to tie together k1 and r1 points to it's corresponding
> /// `Affine<Curve>`
> ```

Can you rephrase? It's not clear



_Code quote:_
```XML
/// Data structure to tie together k1 and r1 points to it's corresponding
/// `Affine<Curve>`
```
___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 681 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfkZBf-P5FVdodlZOi:-OBfkZBg-KvP1B3C_HLr:be48ef6) ([raw file](https://github.com/starkware-libs/sequencer/blob/eb91092cf7161316374f56196a0996516cee1935/crates/blockifier/src/execution/native/syscall_handler.rs#L681)):*
> ```XML
> /// `Affine<Curve>`
> #[derive(PartialEq, Clone, Copy)]
> struct Secp256Point<Curve: SWCurveConfig>(Affine<Curve>);
> ```

Please place the struct above the impl blocks



_Code quote:_
```XML
struct Secp256Point<Curve: SWCurveConfig>(Affine<Curve>);
```
___
*[`crates/blockifier/src/execution/syscalls/secp.rs` line 63 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfoWkI3Qf303fQ0OGX:-OBfoWkI3Qf303fQ0OGY:bit48uc) ([raw file](https://github.com/starkware-libs/sequencer/blob/eb91092cf7161316374f56196a0996516cee1935/crates/blockifier/src/execution/syscalls/secp.rs#L63)):*
> ```XML
> 
>     pub fn secp_new(&mut self, request: SecpNewRequest) -> SyscallResult<SecpNewResponse> {
>         let affine = new_affine::<Curve>(request.x, request.y);
> ```

Why not using `?`?



_Suggestion:_
```XML
let affine = new_affine::<Curve>(request.x, request.y)?;
```
___
*[`crates/blockifier/src/execution/syscalls/secp.rs` line 94 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfoUTN-SdvvPzD-ba-:-OBfoUTN-SdvvPzD-ba0:b-zi71d5) ([raw file](https://github.com/starkware-libs/sequencer/blob/df433f1b888dc73b3c5c122cdc7006779aaa45d5/crates/blockifier/src/execution/syscalls/secp.rs#L94)):*
> ```XML
>         }
>         let ec_point = if x.is_zero() && y.is_zero() {
>             short_weierstrass::Affine::<Curve>::identity()
> ```

What about this case?



_Code quote:_
```XML
        let ec_point = if x.is_zero() && y.is_zero() {
            short_weierstrass::Affine::<Curve>::identity()
```

**Review by varex83 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: 1 of 4 files reviewed, 5 unresolved discussions (waiting on @avi-starkware, @meship-starkware, @noaov1, @xrvdg, and @Yoni-Starkware)

___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 673 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfmlR99hFBCKV7rBdu:-OBgDlzz23VLkrweSgjx:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/eb91092cf7161316374f56196a0996516cee1935/crates/blockifier/src/execution/native/syscall_handler.rs#L673)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Please share this code
</blockquote></details>

Done
___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 679 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfmuYW0VBBCGj9xzDZ:-OBgDlUV9ZK9hgrZkhsb:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/eb91092cf7161316374f56196a0996516cee1935/crates/blockifier/src/execution/native/syscall_handler.rs#L679)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Can you rephrase? It's not clear
</blockquote></details>

Done
___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 681 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfkZBf-P5FVdodlZOi:-OBgDktL66Jum1mm0p8C:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/eb91092cf7161316374f56196a0996516cee1935/crates/blockifier/src/execution/native/syscall_handler.rs#L681)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Please place the struct above the impl blocks
</blockquote></details>

Done
___
*[`crates/blockifier/src/execution/syscalls/secp.rs` line 40 at r13](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBaGRf74t_0VFCNyIvF:-OBgDkBs6vTTINIZ1utb:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/075467f2bce9281864e660c2cb349d2fc6cc2cfd/crates/blockifier/src/execution/syscalls/secp.rs#L40)):*
<details><summary><i>Previously, avi-starkware wrote…</i></summary><blockquote>

@ilyalesokhin-starkware 
?
</blockquote></details>

Done
___
*[`crates/blockifier/src/execution/syscalls/secp.rs` line 63 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfoWkI3Qf303fQ0OGX:-OBgD78lBzmU_0SS9C1N:b-hpsbv5) ([raw file](https://github.com/starkware-libs/sequencer/blob/eb91092cf7161316374f56196a0996516cee1935/crates/blockifier/src/execution/syscalls/secp.rs#L63)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Why not using `?`?
</blockquote></details>

Since we return the same type, and for me, it's easier to just map instead of unwrapping and wrapping into `Ok(..)`
___
*[`crates/blockifier/src/execution/syscalls/secp.rs` line 94 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfoUTN-SdvvPzD-ba-:-OBgDeqA8YQkyKSh6l-D:bvp4l92) ([raw file](https://github.com/starkware-libs/sequencer/blob/df433f1b888dc73b3c5c122cdc7006779aaa45d5/crates/blockifier/src/execution/syscalls/secp.rs#L94)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

What about this case?
</blockquote></details>

It's moved to here:



_Code snippet:_
```XML
fn maybe_affine<Curve: SWCurveConfig>(
    x: Curve::BaseField,
    y: Curve::BaseField,
) -> Option<Affine<Curve>> {
    let ec_point = if x.is_zero() && y.is_zero() {
        Affine::<Curve>::identity()
    } else {
        Affine::<Curve>::new_unchecked(x, y)
    };

    if ec_point.is_on_curve() && ec_point.is_in_correct_subgroup_assuming_on_curve() {
        Some(ec_point)
    } else {
        None
    }
}
```

**Review by rodrigo-pino — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: 1 of 4 files reviewed, 4 unresolved discussions (waiting on @avi-starkware, @meship-starkware, @noaov1, and @Yoni-Starkware)

___
*[`crates/blockifier/src/execution/syscalls/secp.rs` line 63 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBfoWkI3Qf303fQ0OGX:-OBgWKWD4afpcDxcASVs:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/eb91092cf7161316374f56196a0996516cee1935/crates/blockifier/src/execution/syscalls/secp.rs#L63)):*
<details><summary><i>Previously, varex83 (Bohdan Ohorodnii) wrote…</i></summary><blockquote>

Since we return the same type, and for me, it's easier to just map instead of unwrapping and wrapping into `Ok(..)`
</blockquote></details>

Done

**Review by Yoni-Starkware — APPROVED** [review]
:lgtm:

> Reviewed 1 of 4 files at r6, 1 of 3 files at r16, 1 of 1 files at r17, 1 of 1 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-:-OBsuBWkEBu0D2GHL1eX:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @noaov1 and @xrvdg)

___
*[`crates/blockifier/src/execution/secp.rs` line 20 at r18](https://reviewable.io/reviews/starkware-libs/sequencer/1813#-OBssfhlC3OlpF9HlY2Z:-OBssfhlC3OlpF9HlY2_:b-dij4fp) ([raw file](https://github.com/starkware-libs/sequencer/blob/c737b8b8c79066bb9fbdc9b2b669084f96dce5a6/crates/blockifier/src/execution/secp.rs#L20)):*
> ```XML
>         .map(|(smaller, greater)| {
>             // Return the correct y coordinate based on the parity.
>             if ark_ff::BigInteger::is_odd(&smaller.into_bigint()) == y_parity {
> ```

Why not keep it as is?



_Suggestion:_
```XML
smaller.into_bigint().is_odd()
```

**Review by meship-starkware — APPROVED** [review]
> Reviewed 1 of 3 files at r16, 1 of 1 files at r17, 1 of 1 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: all files reviewed, 1 unresolved discussion (waiting on @noaov1 and @xrvdg)

**Review by meship-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: all files reviewed, 1 unresolved discussion (waiting on @noaov1 and @xrvdg)

**Review by Yoni-Starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1813)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @noaov1)
